### PR TITLE
ENH Rework narrative of GBDT notebook

### PR DESCRIPTION
**Pull request by ArturoAmorQ**
Originally I wanted to rework only the wording, as well explaining the GBDT algo is the cornerstone for understanding HGBT. But then I decided to factorize the code by introducing a helper function to keep the focus on the narrative other than the code.

### CODE REVIEW

File: python_scripts/ensemble_gradient_boosting.py [hunk 1]
```diff
@@ -8,26 +8,28 @@
...
+# In this notebook, we present the gradient boosting decision tree algorithm.
 #
-# Gradient-boosting differs from AdaBoost due to the following reason: instead
-# of assigning weights to specific samples, GBDT will fit a decision tree on the
-# residuals error (hence the name "gradient") of the previous tree. Therefore,
-# each new tree in the ensemble predicts the error made by the previous learner
-# instead of predicting the target directly.
+# Even if AdaBoost and GBDT are both boosting algorithms, they are different in
+# nature: the former assigns weights to specific samples, whereas GBDT fits
+# succesive decision trees on the residual errors (hence the name "gradient") of
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:15]:
```suggestion
# successive decision trees on the residual errors (hence the name "gradient") of
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 2]
```diff
@@ -8,26 +8,28 @@
 # %% [markdown]
 # # Gradient-boosting decision tree (GBDT)
 #
-# In this notebook, we will present the gradient boosting decision tree
-# algorithm and contrast it with AdaBoost.
+# In this notebook, we present the gradient boosting decision tree algorithm.
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:11]:
```suggestion
# In this notebook, we present the gradient boosting decision tree (GBDT) algorithm.
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 3]
```diff
@@ -8,26 +8,28 @@
 # %% [markdown]
 # # Gradient-boosting decision tree (GBDT)
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:9]:
```suggestion
# # Gradient-boosting decision tree
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 4]
```diff
@@ -8,26 +8,28 @@
...
+# Later in this notebook we compare the performance of GBDT (boosting) with that
+# of a Random Forest (bagging) for a particular dataset.
 
 # %%
 import pandas as pd
 import numpy as np
 
-# Create a random number generator that will be used to set the randomness
-rng = np.random.RandomState(0)
+rng = np.random.RandomState(0)  # Create a random number generator
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:32]: Let's move the generator next to the data generation. We should avoid showing a pattern where people would use the generator across different function and estimators.
So the best practice is just to show it next to the data generation. we could even slightly change the code and have:

```python
def generate_data(n_samples=50, seed=0):
    rng = np.random.default_rng(seed)
    x = rng.normal(size=(n_samples,)) * ...
    noise = rng.normal(size=(s_samples,)) * 0.3
```

using `default_rng` should be encourage nowadays.

File: python_scripts/ensemble_gradient_boosting.py [hunk 5]
```diff
@@ -74,45 +76,61 @@ def generate_data(n_samples=50):
...
 # %%
-# plot the data
-sns.scatterplot(
-    x=data_train["Feature"], y=target_train, color="black", alpha=0.5
-)
-# plot the predictions
-line_predictions = plt.plot(data_test["Feature"], target_test_predicted, "--")
+def plot_decision_tree_with_residuals(y_train, y_train_pred, y_test_pred):
+    # Create a plot and get the Axes object
+    fig, ax = plt.subplots()
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:91]:
```suggestion
    _fig_, ax = plt.subplots()
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 6]
```diff
@@ -74,45 +76,61 @@ def generate_data(n_samples=50):
...
 
 # %%
-# plot the data
-sns.scatterplot(
-    x=data_train["Feature"], y=target_train, color="black", alpha=0.5
-)
-# plot the predictions
-line_predictions = plt.plot(data_test["Feature"], target_test_predicted, "--")
+def plot_decision_tree_with_residuals(y_train, y_train_pred, y_test_pred):
+    # Create a plot and get the Axes object
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:90]:
```suggestion
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 7]
```diff
@@ -74,45 +76,61 @@ def generate_data(n_samples=50):
...
+
 
 # %%
-# plot the data
-sns.scatterplot(
-    x=data_train["Feature"], y=target_train, color="black", alpha=0.5
-)
-# plot the predictions
-line_predictions = plt.plot(data_test["Feature"], target_test_predicted, "--")
+def plot_decision_tree_with_residuals(y_train, y_train_pred, y_test_pred):
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:89]: Add a proper docstring to this function for good practice.

Comment (glemaitre) [reply python_scripts/ensemble_gradient_boosting.py:89]: It is particularly important to explain what `handles` will be used for.

File: python_scripts/ensemble_gradient_boosting.py [hunk 8]
```diff
@@ -74,45 +76,61 @@ def generate_data(n_samples=50):
...
 
-# plot the residuals
-for value, true, predicted in zip(
-    data_train["Feature"], target_train, target_train_predicted
-):
-    lines_residuals = plt.plot([value, value], [true, predicted], color="red")
 
-plt.legend(
-    [line_predictions[0], lines_residuals[0]], ["Fitted tree", "Residuals"]
+handles, ax = plot_decision_tree_with_residuals(
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:112]:
```suggestion
# %%
handles, ax = plot_decision_tree_with_residuals(
```

File: python_scripts/ensemble_gradient_boosting.py [hunk 9]
```diff
@@ -126,29 +144,22 @@ def generate_data(n_samples=50):
...
-    lines_residuals = plt.plot([value, value], [true, predicted], color="red")
-
-plt.legend(
-    [line_predictions[0], lines_residuals[0]],
-    ["Fitted tree", "Residuals"],
-    bbox_to_anchor=(1.05, 0.8),
-    loc="upper left",
-)
-_ = plt.title("Prediction of the previous residuals")
+legend_labels = [
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:152]: can we pass the legend labels and title as parameter as well like this the `plot_` function takes really care about all plotting.

Comment (glemaitre) [reply python_scripts/ensemble_gradient_boosting.py:152]: And we don't need to define the story about the `handles`

File: python_scripts/ensemble_gradient_boosting.py [hunk 10]
```diff
@@ -265,7 +259,9 @@ def generate_data(n_samples=50):
 # second tree corrects the first tree's error, while the third tree corrects the
 # second tree's error and so on).
 #
-# We will compare the generalization performance of random-forest and gradient
+# ## First comparison of GBDT vs random forests
```
Comment (glemaitre) [inline python_scripts/ensemble_gradient_boosting.py:262]:
```suggestion
# ## First comparison of GBDT vs. random forests
```

Comment (glemaitre) [reply python_scripts/ensemble_gradient_boosting.py:262]: I'm wondering why "First" here. I would drop it since you might refer to another document.

**Review by glemaitre — COMMENTED** [review]
Here are some comments.

File: python_scripts/ensemble_gradient_boosting.py (hunk 9 above)
Comment (ArturoAmorQ) [reply python_scripts/ensemble_gradient_boosting.py:153]: I thought that making the helper accept legend labels and titles may be a bit too much of customization.

Comment (glemaitre) [reply python_scripts/ensemble_gradient_boosting.py:152]: Fair enough.

**Pull request merged**